src/v/serde/parquet/column_writer.cc

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "serde/parquet/column_writer.h"
+
+#include "serde/parquet/encoding.h"
+
+#include <seastar/util/variant_utils.hh>
+
+#include <absl/crc/crc32c.h>
+
+#include <limits>
+#include <stdexcept>
+#include <type_traits>
+
+namespace serde::parquet {
+
+class column_writer::impl {
+public:
+    impl() = default;
+    impl(const impl&) = delete;
+    impl& operator=(const impl&) = delete;
+    impl(impl&&) = default;
+    impl& operator=(impl&&) = default;
+    virtual ~impl() = default;
+
+    virtual void add(value val, uint8_t rep_level, uint8_t def_level) = 0;
+    virtual data_page flush_page() = 0;
+};
+
+namespace {
+
+absl::crc32c_t extend_crc32(absl::crc32c_t crc, const iobuf& buf) {
+    for (const auto& frag : buf) {
+        crc = absl::ExtendCrc32c(crc, {frag.get(), frag.size()});
+    }
+    return crc;
+}
+
+template<typename... Args>
+absl::crc32c_t compute_crc32(Args&&... args) {
+    absl::crc32c_t crc{};
+    ((crc = extend_crc32(crc, std::forward<Args>(args))), ...);
+    return crc;
+}
+
+template<typename value_type>
+class buffered_column_writer final : public column_writer::impl {
+public:
+    buffered_column_writer(int16_t max_def_level, int16_t max_rep_level)
+      : _max_rep_level(max_rep_level)
+      , _max_def_level(max_def_level) {}
+
+    void add(value val, uint8_t rep_level, uint8_t def_level) override {
```

Comment:
nit: is there a reason to use uint8_t instead of rep_level (int16_t)? (same for def_level)